demos/https/iot_demo_https_common.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * Amazon FreeRTOS V201906.00 Major
+ * Copyright (C) 2019 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/**
+ * @file iot_demo_https_common.c
+ * @brief Implementation of common routines used in the HTTPS S3 access demos.
+ */
+/* The config header is always included first. */
+#include "iot_config.h"
+
+/* C Standard headers. */
+#include <string.h>
+
+/* Common header. */
+#include "iot_demo_https_common.h"
+
+/* Amazon FreeRTOS includes. */
+#include "iot_https_utils.h"
+#include "iot_demo_logging.h"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The size of the header value string for the Range: field.
+ *
+ * This is used to specify which parts of the file
+ * we want to download. Let's say the maximum file size is what can fit in a 32 bit unsigned integer. 2^32 = 4294967296
+ * which is 10 digits. The header value string is of the form: "bytes=N-M" where N and M are integers. So the length
+ * of this string is strlen(N) + strlen(M) + strlen("bytes=-") + NULL terminator. Given the maximum number of digits is
+ * 10 we get the maximum length of this header value as: 10 * 2 + 7 + 1.
+ */
+#define RANGE_VALUE_MAX_LENGTH               ( 28 )
+
+/**
+ * @brief HTTP standard header field "Range".
+ */
+#define RANGE_HEADER_FIELD                   "Range"
+#define RANGE_HEADER_FIELD_LENGTH            ( 5 )
+
+/**
+ * @brief HTTP standard header value for requesting a range of bytes from 0 to 0.
+ *
+ * This is used to get the size of the file from S3. Performing a HEAD request with S3 requires generating a Sigv4
```

Comment:
But not for a pre-signed URL right?

Reply:
Yea can't do a HEAD request with a pre-signed URL.

Reply:
Does this make things more clear?
 * This is used to get the size of the file from S3. A HEAD request is forbidden with a presigned URL. Performing a 
 * HEAD request with S3 requires generating a Sigv4 signature in an Authorization header field. We work around this by 
 * performing a GET on Range: bytes=0-0. Then extracting the size of the file from the Content-Range header field in 
 * the response. */